### ISSUE TITLE
[Bug] Using initrd is causing memory corruption

### ISSUE THREAD

**Issue by jeromegn** [firecracker-microvm/firecracker#2325]
## Describe the bug

We've recently switched to using `initrd_path` and noticed memory corruption (at least, that's the theory) about 20% of the time at boot.

We're unsure this is memory corruption, but it consistently shows different problems on boot which led us to this theory.

We built our own `init` program that essentially does this:
- mount /dev
- mkdir /newroot
- mount /dev/vda /newroot
- mount --move /dev /dev/newroot
- chdir /newroot
- mount . /
- chroot .
- chdir /
- ... a few things with /etc/hosts, /etc/resolv.conf, other mounts, etc.
- run a user-specified program

**Switching back to not-using-initrd fixes this issue**. I reused the same exact `init` program, but instead of creating an initrd cpio archive, I just made a block device mounted on /dev/vda.

Running this an init ramdisk produced [as outlined in the docs](https://github.com/firecracker-microvm/firecracker/blob/master/docs/initrd.md) (written by me) shows odd behaviour for the last step. Here are a few examples:

```
dig: symbol lookup error: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: undefined symbol: , version GLIBCXX_3.4.15
[    0.301729] process 'usr/local/bin/keydb-server' started with executable stack
keydb-server: symbol lookup error: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: undefined symbol: , version GLIBCXX_3.4.15
[2020-12-07T17:07:21Z] Program exited with code: 127
```

```
// about 100 lines of this gibberish
���������������������������������������������������������������������
   �1����A��1�A��/7��E��A���D$A����O�D��D��E����A��F�L�: cannot open shared object file: Error 36
```

```
2020-12-07T07:11:32Z [info] [2020-12-07T07:11:32Z] Running: `deno run --allow-net=0.0.0.0,suggestqueries.google.com --allow-env --no-check index.js` as root
2020-12-07T07:11:32Z [info] [2020-12-07T07:11:32Z] Program exited with signal: SIGTRAP (5)
```

```
[  0.35069] dig[503]: segfault at 7f9fd3b0bbcc ip 00007f9f0ca4911a sp 00007ffdd98c7830 error 4 in ld-2.27.so[7f9f0ca3e000+27000]
[   0.352128] Code: 00 00 bf 7f 00 00 00 e8 64 1f 01 00 0f 1f 40 00 41 57 41 56 49 89 f6 41 55 41 54 49 89 d4 55 53 4d89 c5 48 81 ec 98 00 00 00 <0f> b6 17 48 89 3c 24 48 89 4c 24 20 44 89 4c 24 0c 48 8b ac 24 d8
Sementation fault
[    0.355527] process 'usr/local/bin/keydb-server' started with exectable stack
[  ] keydb-server[494]: segfault at 7fdbb697abcc ip 00007fdaeb00311a sp 00007e0370f0 error 4 in ld-2.27.so[7fdaeaff8000+27000]
0] Code: 00 0 bf 7f 00 00 00 e8 64 1f 01 00 0f 1f 40 00 41 57 41 56 49 89 f6 41 55 41 54 49 89 d4 55 53 4d 89 c5 48 81 ec 98 00 00 00 <0f> b6 17 48 89 3c 24 48 89 4c 24 20 44 89 4c 24 0c 48 8b ac 24 d8
[2020-12-07T16:03:02Z] Program exited with signal: SIGSEGV (11)
```

```
Segmentation fault
[2020-12-07T15:59:40Z] Program exited with signal: SIGSEGV (11)
```

<details><summary><strong>Full logs for the first example</strong></summary>

```
[    0.000000] Linux version 5.9.11 (root@34fe72bffe4c) (gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516, GNU ld (GNU Binutils for Debian) 2.28) #1 SMP Tue Nov 24 20:10:27 UTC 2020
[    0.000000] Command line: pci=off LOG_FILTER=init=debug console=ttyS0 reboot=k swapaccount=1 i8042.nomux acpi=off panic=1 rdinit=/fly/init cgroup_enable=memory random.trust_cpu=on i8042.noaux i8042.nopnp i8042.dumbkbd debug virtio_mmio.device=4K@0xd0000000:5 virtio_mmio.device=4K@0xd0001000:6 virtio_mmio.device=4K@0xd0002000:7
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x008: 'MPX bounds registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x010: 'MPX CSR'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
[    0.000000] x86/fpu: xstate_offset[3]:  832, xstate_sizes[3]:   64
[    0.000000] x86/fpu: xstate_offset[4]:  896, xstate_sizes[4]:   64
[    0.000000] x86/fpu: Enabled xstate features 0x1f, context size is 960 bytes, using 'compacted' format.
[    0.000000] BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x0000000007ffffff] usable
[    0.000000] NX (Execute Disable) protection: active
[    0.000000] DMI not present or invalid.
[    0.000000] Hypervisor detected: KVM
[    0.000000] kvm-clock: Using msrs 4b564d01 and 4b564d00
[    0.000000] kvm-clock: cpu 0, msr 2401001, primary cpu clock
[    0.000000] kvm-clock: using sched offset of 29108883 cycles
[    0.000007] clocksource: kvm-clock: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000022] tsc: Detected 2111.998 MHz processor
[    0.000155] e820: update [mem 0x00000000-0x00000fff] usable ==> reserved
[    0.000157] e820: remove [mem 0x000a0000-0x000fffff] usable
[    0.000160] last_pfn = 0x8000 max_arch_pfn = 0x400000000
[    0.000201] MTRR default type: uncachable
[    0.000202] MTRR fixed ranges disabled:
[    0.000204]   00000-FFFFF uncachable
[    0.000205] MTRR variable ranges disabled:
[    0.000206]   0 disabled
[    0.000207]   1 disabled
[    0.000207]   2 disabled
[    0.000208]   3 disabled
[    0.000209]   4 disabled
[    0.000209]   5 disabled
[    0.000210]   6 disabled
[    0.000210]   7 disabled
[    0.000211] Disabled
[    0.000215] x86/PAT: MTRRs disabled, skipping PAT initialization too.
[    0.000225] CPU MTRRs all blank - virtualized system.
[    0.000230] x86/PAT: Configuration [0-7]: WB  WT  UC- UC  WB  WT  UC- UC
[    0.000278] found SMP MP-table at [mem 0x0009fc00-0x0009fc0f]
[    0.000354] check: Scanning 1 areas for low memory corruption
[    0.000370] Using GB pages for direct mapping
[    0.000521] RAMDISK: [mem 0x0798e000-0x07ffffff]
[    0.000647] No NUMA configuration found
[    0.000648] Faking a node at [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000659] NODE_DATA(0) allocated [mem 0x0796c000-0x0798dfff]
[    0.000826] Zone ranges:
[    0.000830]   DMA      [mem 0x0000000000001000-0x0000000000ffffff]
[    0.000831]   DMA32    [mem 0x0000000001000000-0x0000000007ffffff]
[    0.000832]   Normal   empty
[    0.000835] Movable zone start for each node
[    0.000839] Early memory node ranges
[    0.000840]   node   0: [mem 0x0000000000001000-0x000000000009efff]
[    0.000841]   node   0: [mem 0x0000000000100000-0x0000000007ffffff]
[    0.000860] Zeroed struct page in unavailable ranges: 98 pages
[    0.000862] Initmem setup node 0 [mem 0x0000000000001000-0x0000000007ffffff]
[    0.000864] On node 0 totalpages: 32670
[    0.000868]   DMA zone: 64 pages used for memmap
[    0.000869]   DMA zone: 21 pages reserved
[    0.000870]   DMA zone: 3998 pages, LIFO batch:0
[    0.001083]   DMA32 zone: 448 pages used for memmap
[    0.001084]   DMA32 zone: 28672 pages, LIFO batch:7
[    0.002934] Intel MultiProcessor Specification v1.4
[    0.002939] MPTABLE: OEM ID: FC
[    0.002940] MPTABLE: Product ID: 000000000000
[    0.002941] MPTABLE: APIC at: 0xFEE00000
[    0.002969] Processor #0 (Bootup-CPU)
[    0.002998] IOAPIC[0]: apic_id 2, version 17, address 0xfec00000, GSI 0-23
[    0.003003] Processors: 1
[    0.003004] TSC deadline timer available
[    0.003009] smpboot: Allowing 1 CPUs, 0 hotplug CPUs
[    0.003022] kvm-guest: KVM setup pv remote TLB flush
[    0.003040] kvm-guest: setup PV sched yield
[    0.003053] PM: hibernation: Registered nosave memory: [mem 0x00000000-0x00000fff]
[    0.003055] PM: hibernation: Registered nosave memory: [mem 0x0009f000-0x000fffff]
[    0.003057] [mem 0x08000000-0xffffffff] available for PCI devices
[    0.003058] Booting paravirtualized kernel on KVM
[    0.003062] clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645519600211568 ns
[    0.003067] setup_percpu: NR_CPUS:128 nr_cpumask_bits:128 nr_cpu_ids:1 nr_node_ids:1
[    0.005250] percpu: Embedded 44 pages/cpu s143360 r8192 d28672 u2097152
[    0.005261] pcpu-alloc: s143360 r8192 d28672 u2097152 alloc=1*2097152
[    0.005262] pcpu-alloc: [0] 0
[    0.005306] kvm-guest: KVM setup async PF for cpu 0
[    0.005312] kvm-guest: stealtime: cpu 0, msr 7422080
[    0.005316] kvm-guest: PV spinlocks disabled, single CPU
[    0.005330] Built 1 zonelists, mobility grouping on.  Total pages: 32137
[    0.005332] Policy zone: DMA32
[    0.005334] Kernel command line: pci=off LOG_FILTER=init=debug console=ttyS0 reboot=k swapaccount=1 i8042.nomux acpi=off panic=1 rdinit=/fly/init cgroup_enable=memory random.trust_cpu=on i8042.noaux i8042.nopnp i8042.dumbkbd debug virtio_mmio.device=4K@0xd0000000:5 virtio_mmio.device=4K@0xd0001000:6 virtio_mmio.device=4K@0xd0002000:7
[    0.005695] Dentry cache hash table entries: 16384 (order: 5, 131072 bytes, linear)
[    0.005750] Inode-cache hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.005803] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.008031] Memory: 98588K/130680K available (10243K kernel code, 644K rwdata, 1924K rodata, 1432K init, 6008K bss, 31836K reserved, 0K cma-reserved)
[    0.008106] random: get_random_u64 called from __kmem_cache_create+0x2d/0x450 with crng_init=0
[    0.008355] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.009307] rcu: Hierarchical RCU implementation.
[    0.009320] rcu:     RCU restricting CPUs from NR_CPUS=128 to nr_cpu_ids=1.
[    0.009325] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.009326] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=1
[    0.009418] NR_IRQS: 4352, nr_irqs: 48, preallocated irqs: 16
[    0.010116] random: crng done (trusting CPU's manufacturer)
[    0.010393] Console: colour dummy device 80x25
[    0.110892] printk: console [ttyS0] enabled
[    0.111447] APIC: Switch to symmetric I/O mode setup
[    0.112045] x2apic enabled
[    0.112536] Switched APIC routing to physical x2apic.
[    0.112991] kvm-guest: setup PV IPIs
[    0.114015] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x1e71768ef8b, max_idle_ns: 440795218977 ns
[    0.114910] Calibrating delay loop (skipped) preset value.. 4223.99 BogoMIPS (lpj=8447992)
[    0.115954] pid_max: default: 32768 minimum: 301
[    0.116761] LSM: Security Framework initializing
[    0.117626] SELinux:  Initializing.
[    0.118450] Mount-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.118906] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.118906] x86/cpu: User Mode Instruction Prevention (UMIP) activated
[    0.118906] Last level iTLB entries: 4KB 64, 2MB 8, 4MB 8
[    0.118906] Last level dTLB entries: 4KB 64, 2MB 0, 4MB 0, 1GB 4
[    0.118906] Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    0.118906] Spectre V2 : Mitigation: Enhanced IBRS
[    0.118906] Spectre V2 : Spectre v2 / SpectreRSB mitigation: Filling RSB on context switch
[    0.118906] Spectre V2 : mitigation: Enabling conditional Indirect Branch Prediction Barrier
[    0.118906] Speculative Store Bypass: Mitigation: Speculative Store Bypass disabled via prctl and seccomp
[    0.118906] SRBDS: Mitigation: TSX disabled
[    0.118906] Freeing SMP alternatives memory: 32K
[    0.118906] smpboot: CPU0: Intel(R) Xeon(R) Processor @ 1.90GHz (family: 0x6, model: 0x8e, stepping: 0xc)
[    0.118906] Performance Events: unsupported p6 CPU model 142 no PMU driver, software events only.
[    0.118906] rcu: Hierarchical SRCU implementation.
[    0.118906] smp: Bringing up secondary CPUs ...
[    0.118906] smp: Brought up 1 node, 1 CPU
[    0.118915] smpboot: Max logical packages: 1
[    0.119483] smpboot: Total of 1 processors activated (4223.99 BogoMIPS)
[    0.120574] devtmpfs: initialized
[    0.121060] x86/mm: Memory block size: 128MB
[    0.121863] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.122737] futex hash table entries: 256 (order: 2, 16384 bytes, linear)
[    0.123372] NET: Registered protocol family 16
[    0.124357] DMA: preallocated 128 KiB GFP_KERNEL pool for atomic allocations
[    0.125919] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.127044] DMA: preallocated 128 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.128021] audit: initializing netlink subsys (disabled)
[    0.128834] thermal_sys: Registered thermal governor 'fair_share'
[    0.128835] thermal_sys: Registered thermal governor 'step_wise'
[    0.129665] thermal_sys: Registered thermal governor 'user_space'
[    0.130465] cpuidle: using governor ladder
[    0.131386] cpuidle: using governor menu
[    0.137138] audit: type=2000 audit(1607360841.785:1): state=initialized audit_enabled=0 res=1
[    0.140309] HugeTLB registered 1.00 GiB page size, pre-allocated 0 pages
[    0.141167] HugeTLB registered 2.00 MiB page size, pre-allocated 0 pages
[    0.142513] ACPI: Interpreter disabled.
[    0.142994] SCSI subsystem initialized
[    0.143471] pps_core: LinuxPPS API ver. 1 registered
[    0.143969] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.144831] PTP clock support registered
[    0.145265] NetLabel: Initializing
[    0.145556] NetLabel:  domain hash size = 128
[    0.145952] NetLabel:  protocols = UNLABELED CIPSOv4 CALIPSO
[    0.146492] NetLabel:  unlabeled traffic allowed by default
[    0.147390] clocksource: Switched to clocksource kvm-clock
[    0.147978] VFS: Disk quotas dquot_6.6.0
[    0.148322] VFS: Dquot-cache hash table entries: 512 (order 0, 4096 bytes)
[    0.148961] pnp: PnP ACPI: disabled
[    0.151586] NET: Registered protocol family 2
[    0.152060] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.152785] TCP established hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.153483] TCP bind hash table entries: 1024 (order: 2, 16384 bytes, linear)
[    0.154129] TCP: Hash tables configured (established 1024 bind 1024)
[    0.154720] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.155323] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.155950] NET: Registered protocol family 1
[    0.157209] Trying to unpack rootfs image as initramfs...
[    0.163887] Freeing initrd memory: 6600K
[    0.164325] virtio-mmio: Registering device virtio-mmio.0 at 0xd0000000-0xd0000fff, IRQ 5.
[    0.165015] virtio-mmio: Registering device virtio-mmio.1 at 0xd0001000-0xd0001fff, IRQ 6.
[    0.165679] virtio-mmio: Registering device virtio-mmio.2 at 0xd0002000-0xd0002fff, IRQ 7.
[    0.166366] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x1e71768ef8b, max_idle_ns: 440795218977 ns
[    0.167232] clocksource: Switched to clocksource tsc
[    0.167637] platform rtc_cmos: registered platform RTC device (no PNP device found)
[    0.168294] check: Scanning for low memory corruption every 60 seconds
[    0.169094] Initialise system trusted keyrings
[    0.169426] Key type blacklist registered
[    0.169861] workingset: timestamp_bits=36 max_order=15 bucket_order=0
[    0.171004] zbud: loaded
[    0.171337] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.177210] Key type asymmetric registered
[    0.177510] Asymmetric key parser 'x509' registered
[    0.177887] Block layer SCSI generic (bsg) driver version 0.4 loaded (major 252)
[    0.178682] intel_idle: Please enable MWAIT in BIOS SETUP
[    0.179378] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.180028] serial8250: ttyS0 at I/O 0x3f8 (irq = 4, base_baud = 115200) is a 16550A
[    0.181542] loop: module loaded
[    0.182013] virtio_blk virtio0: [vda] 16777216 512-byte logical blocks (8.59 GB/8.00 GiB)
[    0.182630] vda: detected capacity change from 0 to 8589934592
[    0.183578] Loading iSCSI transport class v2.0-870.
[    0.184336] iscsi: registered transport (tcp)
[    0.184813] wireguard: WireGuard 1.0.0 loaded. See www.wireguard.com for information.
[    0.185469] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.186369] tun: Universal TUN/TAP device driver, 1.6
[    0.187223] i8042: PNP detection disabled
[    0.187721] serio: i8042 KBD port at 0x60,0x64 irq 1
[    0.188246] input: AT Raw Set 2 keyboard as /devices/platform/i8042/serio0/input/input0
[    0.189044] intel_pstate: CPU model not supported
[    0.189414] hid: raw HID events driver (C) Jiri Kosina
[    0.189970] Initializing XFRM netlink socket
[    0.190408] NET: Registered protocol family 10
[    0.191254] Segment Routing with IPv6
[    0.191576] NET: Registered protocol family 17
[    0.191963] Bridge firewalling registered
[    0.192329] NET: Registered protocol family 40
[    0.192940] IPI shorthand broadcast: enabled
[    0.193284] sched_clock: Marking stable (86013489, 104899538)->(215739238, -24826211)
[    0.194038] registered taskstats version 1
[    0.194413] Loading compiled-in X.509 certificates
[    0.195420] Loaded X.509 cert 'Build time autogenerated kernel key: ad572c6c113f3d554b0a0e685e36c4959e71a0ea'
[    0.196235] zswap: loaded using pool lzo/zbud
[    0.196650] Key type ._fscrypt registered
[    0.196977] Key type .fscrypt registered
[    0.197361] Key type fscrypt-provisioning registered
[    0.197868] Key type encrypted registered
[    0.198718] Freeing unused decrypted memory: 2040K
[    0.199416] Freeing unused kernel image (initmem) memory: 1432K
[    0.203049] Write protecting the kernel read-only data: 14336k
[    0.206384] Freeing unused kernel image (text/rodata gap) memory: 2044K
[    0.207597] Freeing unused kernel image (rodata/data gap) memory: 124K
[    0.208605] Run /fly/init as init process
[    0.209178]   with arguments:
[    0.209557]     /fly/init
[    0.209783]   with environment:
[    0.210184]     HOME=/
[    0.210470]     TERM=linux
[    0.210886]     LOG_FILTER=init=debug
[    0.211446]     cgroup_enable=memory
[2020-12-07T17:07:21Z] Starting init (commit: 7e8393b)...
[2020-12-07T17:07:21Z] Mounting /dev
[2020-12-07T17:07:21Z] Mounting newroot fs
[    0.215253] EXT4-fs (vda): mounted filesystem with ordered data mode. Opts:
[2020-12-07T17:07:21Z] Mounting (move) /dev
[2020-12-07T17:07:21Z] Moving (copying) /fly
[2020-12-07T17:07:21Z] Switching root
[2020-12-07T17:07:21Z] Mounting /dev/pts
[2020-12-07T17:07:21Z] Mounting /dev/mqueue
[2020-12-07T17:07:21Z] Mounting /dev/shm
[2020-12-07T17:07:21Z] Mounting /dev/hugepages
[2020-12-07T17:07:21Z] Mounting /proc
[2020-12-07T17:07:21Z] Mounting /sys
[2020-12-07T17:07:21Z] Mounting /run
[2020-12-07T17:07:21Z] Mounting cgroup
[2020-12-07T17:07:21Z] Mounting cgroup2
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/net_cls,net_prio
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/hugetlb
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/pids
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/freezer
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/cpu,cpuacct
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/devices
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/blkio
[2020-12-07T17:07:21Z] Mounting cgroup/memory
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/perf_event
[2020-12-07T17:07:21Z] Mounting /sys/fs/cgroup/cpuset
[2020-12-07T17:07:21Z] searching for user 'root
[2020-12-07T17:07:21Z] Populating /etc/hosts
[2020-12-07T17:07:21Z] Populating /etc/resolv.conf
[2020-12-07T17:07:21Z] netlink: getting lo link
[2020-12-07T17:07:21Z] netlink: setting lo link "up"
[2020-12-07T17:07:21Z] netlink: getting eth0 link
[2020-12-07T17:07:21Z] netlink: setting eth0 link "up"
[2020-12-07T17:07:21Z] netlink: adding ip 172.19.1.74/29
[2020-12-07T17:07:21Z] netlink: adding default route via 172.19.1.73/32
[2020-12-07T17:07:21Z] netlink: adding ip 2001:470:1c:3a1:0:8d99:e91e:1/127
[2020-12-07T17:07:21Z] netlink: adding default route via 2001:470:1c:3a1:0:8d99:e91e:0/128
[2020-12-07T17:07:21Z] Running: `/fly/start_keydb.sh` as root
[2020-12-07T17:07:21Z] Setting current dir on command to: /data
[2020-12-07T17:07:21Z] child pid: 494
vm.overcommit_memory = 1
net.core.somaxconn = 1024
dig: symbol lookup error: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: undefined symbol: , version GLIBCXX_3.4.15
[    0.301729] process 'usr/local/bin/keydb-server' started with executable stack
keydb-server: symbol lookup error: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: undefined symbol: , version GLIBCXX_3.4.15
[2020-12-07T17:07:21Z] Program exited with code: 127
[2020-12-07T17:07:22Z] exiting after delay
[    1.337197] kvm-guest: Unregister pv shared memory for cpu 0
[    1.339142] reboot: Restarting system
[    1.340068] reboot: machine restart
```

</details>

## To Reproduce

- Run firecracker with an init ramdisk
- Not entirely sure what else leads to this issue unfortunately

## Expected behaviour

I expected creating the firecrackers with the same configuration would work all the time, even when using an initrd.

## Environment

- Firecracker version: v0.23.0
- Host kernel: 5.9.12 (arch) and 5.6.0
- Guest kernel: 5.9.11
- Rootfs used: various, unpacked from docker images. Some full-blown operating systems, others scratch images.
- Architecture: AMD and Intel

## Checks

- [x] Have you searched the Firecracker Issues database for similar problems?
- [x] Have you read the existing relevant Firecracker documentation?
- [ ] Are you certain the bug being reported is a Firecracker issue? not certain, but I don't see what else could be causing this.

**Comment by georgepisaltu** [comment]
The investigation so far has revealed that it is in fact a binary in the guest that is crashing, not Firecracker itself. The possibility of memory corruption seems to be only inside the guest and is limited to contents of the file system, possibly the shared libraries, and it seems to be happening during the mounting of the rootfs from the initrd. The root cause is still unknown and we have not managed to reproduce it yet.

**Comment by jeromegn** [comment]
It is indeed a binary in the guest rootfs that's crashing, but it's different ones for different reasons, at different times.

I've been able to reproduce this consistently. I'm not sure what's causing it yet.

It appears timing is everything. For example, running the VM with the kernel arg `debug` doesn't trigger the issue, but removing it and/or using `quiet` (therefore, printing less to the console) triggers it. I've rewritten our initrd about 10 times and each produced sightly different results. I believe because the timings changed a little bit every time.

Here's our ramdisk's `init` code:

```rust
fn pre_init() -> Result<(), InitError> {
    info!("Starting pre-init");

    // can't put these as const unfortunately...
    let chmod_0755: Mode =
        Mode::S_IRWXU | Mode::S_IRGRP | Mode::S_IXGRP | Mode::S_IROTH | Mode::S_IXOTH;
    let chmod_0666: Mode = Mode::S_IRUSR
        | Mode::S_IWUSR
        | Mode::S_IRGRP
        | Mode::S_IWGRP
        | Mode::S_IROTH
        | Mode::S_IWOTH;
    {
        let conf: RunConfig = {
            let reader = BufReader::new(File::open("/fly/run.json").map_err(ConfigError::from)?);
            serde_json::from_reader(reader).map_err(ConfigError::from)?
        };

        trace!("run conf: {:#?}", conf);

        debug!("Mounting /dev");
        mkdir("/dev", chmod_0755).ok();
        mount(
            Some("devtmpfs"),
            "/dev",
            Some("devtmpfs"),
            MsFlags::MS_NOSUID,
            Some("mode=0755"),
        )?;

        mkdir("/newroot", chmod_0755)?;

        let root_device = if let Some(rd) = conf.root_device {
            rd
        } else {
            "/dev/vdb".to_owned()
        };

        debug!("Mounting newroot fs");
        mount::<_, _, _, [u8]>(
            Some(root_device.as_str()),
            "/newroot",
            Some("ext4"),
            MsFlags::MS_RELATIME,
            None,
        )?;

        debug!("Changing dir to /newroot");
        chdir("/newroot")?;

        // Move /dev so we don't have to re-mount it
        debug!("Mounting (move) /dev");
        mkdir("/newroot/dev", chmod_0755).ok();
        mount::<_, _, [u8], [u8]>(Some("/dev"), "/newroot/dev", None, MsFlags::MS_MOVE, None)?;
        // debug!("Unmounting /dev");
        // umount("/dev")?;

        debug!("Moving /fly to /newroot/fly");
        mkdir("/newroot/fly", chmod_0755).ok();
        fs::copy("/fly/init", "/newroot/fly/init")?;
        fs::copy("/fly/run.json", "/newroot/fly/run.json")?;

        let root_dev = nix::sys::stat::stat("/")?.st_dev;

        debug!("Deleting contents of / (preparing for overmount)");
        delete_contents("/", root_dev)?;

        // Our own hacky switch_root
        debug!("Switching root");
        // move /newroot (.) to /
        mount::<_, _, [u8], [u8]>(Some("."), "/", None, MsFlags::MS_MOVE, None)?;
        // Change root to the current directory (new root)
        chroot(".")?;
        // Change directory to /
        chdir("/")?;

        match nix::fcntl::open("/dev/console", nix::fcntl::OFlag::O_RDWR, chmod_0666) {
            Ok(rawfd) => {
                trace!("opened /dev/console and got fd: {}", rawfd);
                if rawfd >= 0 {
                    debug!("Moving stdin, stdout and stderr to new root");
                    nix::unistd::dup2(rawfd, 0)?;
                    nix::unistd::close(rawfd)?;
                    nix::unistd::dup2(0, 1)?;
                    nix::unistd::dup2(0, 2)?;
                }
            }
            Err(e) => {
                warn!("could not open /dev/console: {}", e);
            }
        }

        sync();
    }

    let fly_init = CString::new("/fly/init").map_err(Error::from)?;
    let arg0 = CString::new("init").map_err(Error::from)?;
    return nix::unistd::execv(fly_init.as_c_str(), &[arg0.as_c_str()])
        .map(|_| {})
        .map_err(InitError::from);
}

fn delete_contents<P: AsRef<std::path::Path>>(path: P, root_dev: u64) -> Result<(), InitError> {
    use nix::sys::stat::{lstat, SFlag};
    trace!(
        "delete_contents {} (root dev: {})",
        path.as_ref().display(),
        root_dev
    );
    let st = lstat(path.as_ref())?;
    if st.st_dev != root_dev {
        trace!("was not on the root device");
        return Ok(());
    }

    trace!("st.st_mode: {}", st.st_mode);

    if unsafe { SFlag::from_bits_unchecked(st.st_mode & SFlag::S_IFMT.bits()) } == SFlag::S_IFDIR {
        trace!("is a directory");
        if let Ok(dirs) = fs::read_dir(&path) {
            for entry in dirs {
                if let Ok(entry) = entry {
                    let newdir = entry.path();
                    if newdir.as_os_str() == "." || newdir.as_os_str() == ".." {
                        continue;
                    }
                    delete_contents(newdir, root_dev)?;
                }
            }
            trace!("removing directory {}", path.as_ref().display());
            if let Err(e) = fs::remove_dir(&path) {
                trace!(
                    "error removing directory {} => {}",
                    path.as_ref().display(),
                    e
                );
            }
        }
    } else {
        trace!("was not a directory, unlinking!");
        if let Err(e) = nix::unistd::unlink(path.as_ref()) {
            trace!("error unlinking {} => {}", path.as_ref().display(), e);
        }
    }

    Ok(())
}
```

This is mainly using the `nix` crate.

I don't know if this is a firecracker issue in the end, but it's odd that it happens. I think this initrd is fine! It's mostly modeled after https://github.com/brgl/busybox/blob/master/util-linux/switch_root.c and various other posts I've read about it (gentoo wiki, stackoverflow answers, other switch_root implementations, etc.) It's likely I have missed an implementation detail, though.

However, @georgepisaltu has not been able to reproduce without our init. I have only recently supplied them with our init, a rootfs and our vmlinux.

**Comment by jeromegn** [comment]
This is now been fixed in #2403.

Thanks to @georgepisaltu's relentless work over the past few weeks! I'm glad we've been able to work so closely with somebody on this particular issue.

I made sure to test with both stable firecracker and a build off of the latest commit to confirm this was indeed fixed.
